fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java

```diff
@@ -72,4 +90,124 @@ The other damage method (which original.call discards) handles the creative mode
+			@Share("index") LocalIntRef index
+	) {
+		preAppendTooltip(componentType, context, displayComponent, textConsumer, type, index);
+		return componentType;
+	}
+
+	@ModifyArg(method = "appendTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/component/type/TooltipDisplayComponent;shouldDisplay(Lnet/minecraft/component/ComponentType;)Z"))
+	private ComponentType<?> preShouldDisplay(
+			ComponentType<?> componentType,
+			@Local(argsOnly = true) Item.TooltipContext context,
+			@Local(argsOnly = true) TooltipDisplayComponent displayComponent,
+			@Local(argsOnly = true) TooltipType type,
+			@Local(argsOnly = true) Consumer<Text> textConsumer,
+			@Share("index") LocalIntRef index
+	) {
+		preAppendTooltip(componentType, context, displayComponent, textConsumer, type, index);
+		return componentType;
+	}
+
+	@Inject(method = "appendTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;appendAttributeModifiersTooltip(Ljava/util/function/Consumer;Lnet/minecraft/component/type/TooltipDisplayComponent;Lnet/minecraft/entity/player/PlayerEntity;)V"))
+	private void preAttributeModifiers(
+			Item.TooltipContext context,
+			TooltipDisplayComponent displayComponent,
+			@Nullable PlayerEntity player,
+			TooltipType type,
+			Consumer<Text> textConsumer,
+			CallbackInfo ci,
+			@Share("index") LocalIntRef index
+	) {
+		// Special case: attribute modifiers are extracted into a separate method
+		preAppendTooltip(DataComponentTypes.ATTRIBUTE_MODIFIERS, context, displayComponent, textConsumer, type, index);
+	}
+
+	@Inject(method = "appendTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/registry/DefaultedRegistry;getId(Ljava/lang/Object;)Lnet/minecraft/util/Identifier;"))
+	private void postTooltipsAdvanced(
+			Item.TooltipContext context,
+			TooltipDisplayComponent displayComponent,
+			@Nullable PlayerEntity player,
+			TooltipType type,
+			Consumer<Text> textConsumer,
+			CallbackInfo ci,
+			@Share("index") LocalIntRef index
+	) {
+		preAppendTooltip(null, context, displayComponent, textConsumer, type, index);
+	}
+
+	@ModifyExpressionValue(method = "appendTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/tooltip/TooltipType;isAdvanced()Z"))
+	private boolean postTooltipsNonAdvanced(
+			boolean isAdvanced,
+			Item.TooltipContext context,
+			TooltipDisplayComponent displayComponent,
+			@Nullable PlayerEntity player,
+			TooltipType type,
+			Consumer<Text> textConsumer,
+			@Share("index") LocalIntRef index
+	) {
+		if (!isAdvanced) {
+			preAppendTooltip(null, context, displayComponent, textConsumer, type, index);
+		}
+
+		return isAdvanced;
+	}
+
+	@Unique
+	private void preAppendTooltip(
+			@Nullable ComponentType<?> componentType,
+			Item.TooltipContext context,
+			TooltipDisplayComponent displayComponent,
+			Consumer<Text> textConsumer,
+			TooltipType tooltipType,
+			LocalIntRef index
+	) {
+		if (index.get() == 0) {
```

Comment:
Would it be a good idea to skip over all of this logic if no mod has used the registry? Just have something along the lines of `if (ComponentTooltipAppenderRegistryImpl.unmodified()) return;`. Would be easy enough to do, and would save a bit of frame time in vanilla?